File: frontend/src/api/movies.ts
import { request } from "./client";
import type { Movie } from "@/types/movie";

export interface MoviePage {
  movies: Movie[];
  total: number;
}

export interface BrowseParams {
  genre?: string;
  sort?: "popular" | "rating" | "year" | "title";
  query?: string;
  offset?: number;
  limit?: number;
}

/** GET /api/genres */
export async function getGenres(): Promise<string[]> {
  return request<string[]>("/api/genres");
}

/** GET /api/movies?genre=&sort=&query=&offset=&limit= */
export async function browseMovies({
  genre = "All",
  sort = "popular",
  query = "",
  offset = 0,
  limit = 48,
}: BrowseParams = {}): Promise<MoviePage> {
  const params = new URLSearchParams({
    genre,
    sort,
    query,
    offset: String(offset),
    limit: String(limit),
  });
  return request<MoviePage>(`/api/movies?${params.toString()}`);
}

/** GET /api/movies */
export async function getMovies(): Promise<Movie[]> {
  const page = await request<MoviePage>("/api/movies?limit=60");
  return page.movies;
}

/** GET /api/movies/:id */
export async function getMovie(movieId: number): Promise<Movie | null> {
  try {
    return await request<Movie>(`/api/movies/${movieId}`);
  } catch {
    return null;
  }
}

/** GET /api/search?query= */
export async function searchMovies(query: string): Promise<Movie[]> {
  const q = query.trim();
  if (!q) return [];
  return request<Movie[]>(`/api/search?query=${encodeURIComponent(q)}`);
}

export async function getMoviesByIds(ids: number[]): Promise<Movie[]> {
  if (!ids.length) return [];
  const promises = ids.map((id) => getMovie(id));
  const results = await Promise.all(promises);
  return results.filter((m): m is Movie => m !== null);
}

File: backend/app/routers/movies.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session
from sqlalchemy import desc, asc, func, or_, cast, String
from app.database import get_db
from app.models.db_models import Movie
from app.schemas.schemas import MovieSchema, MoviePageSchema
from app.ml.recommender import movie_model_to_schema
from app.ml.tmdb import ensure_movie_posters

router = APIRouter(prefix="/api", tags=["Movies"])

GENRE_LIST = [
    "Action", "Adventure", "Animation", "Children", "Comedy", "Crime",
    "Documentary", "Drama", "Fantasy", "Film-Noir", "Horror", "Musical",
    "Mystery", "Romance", "Sci-Fi", "Thriller", "War", "Western", "IMAX"
]

@router.get("/genres", response_model=List[str])
def get_genres():
    return GENRE_LIST

@router.get("/movies", response_model=MoviePageSchema)
def get_movies(
    offset: int = Query(0, ge=0),
    limit: int = Query(48, ge=1, le=5000),
    genre: str = Query("All"),
    sort: str = Query("popular"),
    query: str = Query(""),
    db: Session = Depends(get_db),
):
    q = db.query(Movie)

    if genre and genre != "All":
        q = q.filter(cast(Movie.genres, String).ilike(f"%{genre}%"))

    if query.strip():
        search_pattern = f"%{query.strip().lower()}%"
        q = q.filter(func.lower(Movie.title).like(search_pattern))

    total = q.count()

    # Sorting
    if sort == "rating":
        q = q.order_by(desc(Movie.rating), desc(Movie.rating_count))
    elif sort == "year":
        q = q.order_by(desc(Movie.year), desc(Movie.rating_count))
    elif sort == "title":
        q = q.order_by(asc(Movie.title))
    else:  # default 'popular'
        q = q.order_by(desc(Movie.rating_count), desc(Movie.rating))

    movies = q.offset(offset).limit(limit).all()
    ensure_movie_posters(movies, db)
    return MoviePageSchema(
        movies=[movie_model_to_schema(m) for m in movies],
        total=total,
    )

@router.get("/movies/{movie_id}", response_model=MovieSchema)
def get_movie_by_id(movie_id: int, db: Session = Depends(get_db)):
    movie = db.query(Movie).filter(Movie.movie_id == movie_id).first()
    if not movie:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Movie with ID {movie_id} not found in catalog",
        )
    ensure_movie_posters([movie], db)
    return movie_model_to_schema(movie)

@router.get("/search", response_model=List[MovieSchema])
def search_movies(
    query: str = Query("", min_length=1),
    limit: int = Query(60, ge=1, le=100),
    db: Session = Depends(get_db),
):
    q_str = query.strip().lower()
    if not q_str:
        return []

    results = (
        db.query(Movie)
        .filter(
            or_(
                func.lower(Movie.title).like(f"%{q_str}%"),
                cast(Movie.genres, String).ilike(f"%{q_str}%"),
            )
        )
        .order_by(
            desc(func.lower(Movie.title) == q_str),
            desc(func.lower(Movie.title).startswith(q_str)),
            desc(Movie.rating_count),
        )
        .limit(limit)
        .all()
    )
    ensure_movie_posters(results, db)
    return [movie_model_to_schema(m) for m in results]
